include/zio/zio_dev.h

```diff
@@ -129,75 +177,76 @@ struct zio_dev_chan {
 	 */
 	const enum byteorder {
 		ZIO_BYTEORDER_ARCH,
-			ZIO_BYTEORDER_BIG,
-			ZIO_BYTEORDER_LITTLE
+		ZIO_BYTEORDER_BIG,
+		ZIO_BYTEORDER_LITTLE
 	} byte_order;
 
 	const enum sign_bit {
 		ZIO_SIGN_NONE,
-			ZIO_SIGN_MSB,
-			ZIO_SIGN_LSB,
+		ZIO_SIGN_MSB,
+		ZIO_SIGN_LSB,
 	} sign_bit;
-
 };
 
+
 /**
- * @brief Function to set a device attribute
+ * @brief Function to set a channel attribute for a device
  *
- * @return -EINVAL if an invalid attribute id was given, 0 otherwise
+ * @return -EINVAL if an invalid attribute id, type, or range was given,
+ * 0 otherwise
  */
-typedef int (*zio_dev_set_attr_t)(struct device *dev, const u32_t attr_idx,
-				  const struct zio_variant var);
+typedef int (*zio_set_attr_t)(struct device *dev, const u16_t channel_idx, const u16_t attribute_idx, const struct zio_variant var);
 
 /**
- * @brief Function to get a device attribute
+ * @brief Function to get a device attribute channel
  *
  * @return -EINVAL if an invalid attribute id was given, 0 otherwise
  */
-typedef int (*zio_dev_get_attr_t)(struct device *dev, uint32_t attr_idx,
-				  struct zio_variant *attr);
+typedef int (*zio_get_attr_t)(struct device *dev, const u16_t channel_idx, const u16_t attribute_idx, struct zio_variant *attr);
+
+
+/** Attribute record. */
+struct zio_attr_desc {
+	/* Attribute data type describing the expected data type. */
+	enum zio_variant_type data_type;
+	zio_set_attr_t set_attr;
```

Comment:
These two function pointers are I think the most significant part of this change set and I definitely agree that they make the device driver API clearer, avoids unnecessary switch/case function wrappers, and generally makes for a nicer API